pkg-config: set pkg-config-rs-compatible env vars

alyssais opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

alyssais committed:
cargo: run libcurl link test when cross-compiling

It's not possible to run installCheckPhase when cross-compiling, so we
have to tack this on to postInstall instead.

alyssais committed:
pkg-config: set pkg-config-rs-compatible env vars

This was already set by the rustc derivation, but by doing it
everywhere pkg-config is used, the necessary environment variables will
also be set for other Rust packages that need them, like cargo.

yu-re-ka commented on pkgs/build-support/pkg-config-wrapper/setup-hook.sh:
```diff
@@ -25,5 +25,11 @@ addEnvHooks "$targetOffset" pkgConfigWrapper_addPkgConfigPath
+# but it does support checking these idiosyncratic PKG_CONFIG_${TRIPLE}
+# environment variables.
+# [1]: https://github.com/rust-lang/pkg-config-rs/issues/53
+export PKG_CONFIG_@rustTarget@=@targetPrefix@@baseBinName@
```
I can't really approve of this rust-specific thing being added in all derivations that use pkg-config, even non-rust ones. I also don't disapprove, but for a quick merge you could move this change out of this PR and let me review only the cargo stuff (with that `PKG_CONFIG_` variable being set just like in rustc).

Mic92 replied:
What if it was in the rustc setuphook? Than as soon as you add rustc to a build, you would get it.

alyssais replied:
I think the pkg-config wrapper is the right place for this.  It's the only place that makes sense that is shared between the rustc build, and all the builds that use Cargo.

Technically, it doesn't have anything to do with pkg-config (since it comes from the pkg-config crate), but it doesn't have anything to do with Cargo either.  It's also extremely similar to PKG_CONFIG_FOR_BUILD, defined on the line above, so it makes sense to set both very similar environment variables in one place.

Ericson2314 replied:
I agree it doesn't go here.

We can have a separate setup hook that Rust. `buildRustCrate`, `buildRustPackage`, etc. all use.

Ericson2314 replied:
https://github.com/rust-lang/pkg-config-rs/issues/139 Also, I opened an issue in the relevant package `build.rs` file use.

alyssais replied:
I don't want more setup hooks that have to be manually added to a build — it makes it too complicated for package authors to do the right thing, especially when it's not obvious they need to do it, like in this case where only cross-compilation is affected.  So what package could bring this in, if not for pkg-config?  cargo?  This env var is even less related to cargo than it is to pkg-config.

alyssais replied:
And what's the justification for letting `PKG_CONFIG_FOR_BUILD` be set here and not the pkg-config-rs environment variables?

`PKG_CONFIG_FOR_BUILD` doesn't come from the pkg-config package either, at least as far as I can tell.  It's an autoconf/Meson/etc thing, so if we were to treat it as you're suggesting we do the pkg-config-rs variables, we'd have to move it into some other setup hook too.

Ericson2314 replied:
See what I left in the upstream stuff, but I really rather have it be a separate setup hook. There are many such things like Cargo/build.rs I don't think it's write to support them all in the tool wrappers. M tools and N consumption patterns I think should be handled separately so one "pays for just what one uses" and complexity is kept under control.

alyssais replied:
I think this is premature.  There are two conventions, and for each one it's setting a single environment variable.  "Paying for what you use" doesn't make sense when the cost is negligible.

Ericson2314 replied:
It's a small amount of code, but changing it is mass rebuild. So I think about the "mass rebuild" costs of these sorts of things too.

Ericson2314 replied:
Also think how if we do two setup hooks it's easy to combine later, but if we do one setup hook it's hard to split later.

alyssais replied:
That would be an extremely rare change to make, especially for the limited set of architectures we actually build for.  We do full rebuilds every couple of weeks anyway, so I don't think it's at all worth optimising for avoidance of mass-rebuilds in infrastructural stuff like this.  Indeed, avoiding mass rebuilds is how we end up with unnecessary conditionals making maintenance more complicated forever, because it avoided mass rebuilds once.  And Rust is so widely used that any change to it is going to cause huge rebuilds anyway, regardless of whether it affects pkg-config.

alyssais replied:
And having a separate setup hook just for this (`cargoPkgConfigSetupHook`?) would mean broken cross packages forever, because since it would only be required for cross, nobody would ever remember to apply it.

yu-re-ka reviewed (approved):
The first time I looked at this, it was not obvious that `PKG_CONFIG_FOR_BUILD` is set directly above the newly added code, and that it is autoconf/meson specific.
I now think this might be fine, but would give @Ericson2314 the last word (to change their mind or explain why this is still a bad idea)

Mindavi commented:
@Ericson2314 and @alyssais, what would be a good way to move this forward? I do kinda agree that it _feels_ wrong to modify the pkg-config wrapper, if a hook were to be added, could that be added to buildRustPackage or injected when cargo or rust is added? Then there might be less of a need to remember to add it.

alyssais commented:
> @Ericson2314 and @alyssais, what would be a good way to move this forward? I do kinda agree that it _feels_ wrong to modify the pkg-config wrapper, if a hook were to be added, could that be added to buildRustPackage or injected when cargo or rust is added? Then there might be less of a need to remember to add it.

I asked John to expand on his most recent comment on Matrix, and AIUI we're still waiting for him to find the time to talk it through with me.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/unable-to-build-pkgsstatic-gcc/27795/6

alyssais committed:
rustc: unset PKG_CONFIG_*

This is no longer required now that this variable is set by the
pkg-config wrapper.

alyssais committed:
cargo: unset HOST_PKG_CONFIG_PATH

This is obsolete now that the pkg-config wrapper sets variables for
Rust triples.

K900 reviewed (approved):
LGTM, we discussed this a bit on Matrix earlier.

emilazy commented:
We may be able to drop `PKG_CONFIG_ALLOW_CROSS` now. It will mean that targets without a `pkg-config` dependency will [fail to use `pkg-config`](https://docs.rs/pkg-config/latest/src/pkg_config/lib.rs.html#606-630) and maybe fall back to non‐`pkg-config` methods, but that’s at least consistent with the non‐cross behaviour. Otherwise it will run plain `pkg-config` when there isn’t a variable for the appropriate target, because we set `PKG_CONFIG_ALLOW_CROSS`, and that seems bad.

The other alternative is setting `PKG_CONFIG_SYSROOT_DIR` per target and using one `pkg-config`, but I think our wrappers don’t work well with that at present?
